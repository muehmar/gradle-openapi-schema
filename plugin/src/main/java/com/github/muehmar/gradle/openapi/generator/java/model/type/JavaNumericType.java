package com.github.muehmar.gradle.openapi.generator.java.model.type;

import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.java.model.name.QualifiedClassName;
import com.github.muehmar.gradle.openapi.generator.java.model.name.QualifiedClassNames;
import com.github.muehmar.gradle.openapi.generator.model.constraints.Constraints;
import com.github.muehmar.gradle.openapi.generator.model.type.NumericType;
import com.github.muehmar.gradle.openapi.generator.settings.FormatTypeMapping;
import com.github.muehmar.gradle.openapi.generator.settings.TypeMappings;
import java.util.EnumMap;
import java.util.Map;
import java.util.Optional;
import java.util.function.Function;
import lombok.EqualsAndHashCode;
import lombok.ToString;

@EqualsAndHashCode(callSuper = true)
@ToString
public class JavaNumericType extends NonGenericJavaType {
  private static final Map<NumericType.Format, QualifiedClassName> FORMAT_CLASS_NAME_MAP =
      createFormatClassNameMap();
  private final Constraints constraints;

  protected JavaNumericType(
      QualifiedClassName className, Constraints constraints, NumericType numericType) {
    super(className, numericType);
    this.constraints = constraints;
  }

  public static JavaNumericType wrap(NumericType numericType, TypeMappings typeMappings) {
    final QualifiedClassName className =
        classNameFromFormat(numericType, typeMappings.getFormatTypeMappings());
    final QualifiedClassName finalClassName =
        className.mapWithClassMappings(typeMappings.getClassTypeMappings());
    return new JavaNumericType(finalClassName, numericType.getConstraints(), numericType);
  }

  private static QualifiedClassName classNameFromFormat(
      NumericType numericType, PList<FormatTypeMapping> formatTypeMappings) {
    final Optional<QualifiedClassName> userFormatMappedClassName =
        QualifiedClassName.fromFormatTypeMapping(
            numericType.getFormat().asString(), formatTypeMappings);
    final QualifiedClassName formatMappedClassName =
        Optional.ofNullable(FORMAT_CLASS_NAME_MAP.get(numericType.getFormat()))
            .orElse(QualifiedClassNames.DOUBLE);
    return userFormatMappedClassName.orElse(formatMappedClassName);
  }

  private static Map<NumericType.Format, QualifiedClassName> createFormatClassNameMap() {
    final Map<NumericType.Format, QualifiedClassName> map = new EnumMap<>(NumericType.Format.class);
    map.put(NumericType.Format.DOUBLE, QualifiedClassNames.DOUBLE);
    map.put(NumericType.Format.FLOAT, QualifiedClassNames.FLOAT);
    return map;
  }

  @Override
  public boolean isJavaArray() {
    return false;
  }

  @Override
  public Constraints getConstraints() {
    return constraints;
  }

  @Override
  public <T> T fold(
      Function<JavaArrayType, T> onArrayType,
      Function<JavaBooleanType, T> onBooleanType,
      Function<JavaEnumType, T> onEnumType,
      Function<JavaMapType, T> onMapType,
      Function<JavaAnyType, T> onAnyType,
      Function<JavaNumericType, T> onNumericType,
      Function<JavaIntegerType, T> onIntegerType,
      Function<JavaObjectType, T> onObjectType,
      Function<JavaStringType, T> onStringType) {
    return onNumericType.apply(this);
  }
}
