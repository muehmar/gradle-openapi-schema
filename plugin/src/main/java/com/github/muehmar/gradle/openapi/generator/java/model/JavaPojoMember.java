package com.github.muehmar.gradle.openapi.generator.java.model;

import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.java.generator.enumpojo.EnumContentBuilder;
import com.github.muehmar.gradle.openapi.generator.java.generator.enumpojo.EnumGenerator.EnumContent;
import com.github.muehmar.gradle.openapi.generator.java.model.name.IsNullFlagName;
import com.github.muehmar.gradle.openapi.generator.java.model.name.IsPresentFlagName;
import com.github.muehmar.gradle.openapi.generator.java.model.name.JavaName;
import com.github.muehmar.gradle.openapi.generator.java.model.type.JavaEnumType;
import com.github.muehmar.gradle.openapi.generator.java.model.type.JavaType;
import com.github.muehmar.gradle.openapi.generator.model.Necessity;
import com.github.muehmar.gradle.openapi.generator.model.Nullability;
import com.github.muehmar.gradle.openapi.generator.model.PojoMember;
import com.github.muehmar.gradle.openapi.generator.model.type.EnumType;
import com.github.muehmar.gradle.openapi.generator.settings.GetterSuffixes;
import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import com.github.muehmar.gradle.openapi.generator.settings.TypeMappings;
import io.github.muehmar.pojobuilder.annotations.PojoBuilder;
import java.util.Optional;
import java.util.function.Function;
import lombok.EqualsAndHashCode;
import lombok.ToString;
import lombok.With;

/**
 * This member corresponds to a property in the specification. To support all possible states af a
 * member, a single {@link JavaPojoMember} may result in more than one {@link TechnicalPojoMember}
 * in a pojo.
 */
@EqualsAndHashCode
@ToString
@PojoBuilder
@With
public class JavaPojoMember {
  private final JavaName name;
  private final String description;
  private final JavaType javaType;
  private final Necessity necessity;
  private final Nullability nullability;
  private final MemberType type;

  private static final String TRISTATE_TO_PROPERTY =
      "onValue(val -> val).onNull(() -> null).onAbsent(() -> null)";
  private static final String TRISTATE_TO_ISNULL_FLAG =
      "onValue(ignore -> false).onNull(() -> true).onAbsent(() -> false)";

  JavaPojoMember(
      JavaName name,
      String description,
      JavaType javaType,
      Necessity necessity,
      Nullability nullability,
      MemberType type) {
    this.javaType = javaType;
    this.name = name;
    this.description = description;
    this.necessity = necessity;
    this.nullability = nullability;
    this.type = type;
  }

  public static JavaPojoMember wrap(PojoMember pojoMember, TypeMappings typeMappings) {
    final JavaType javaType = JavaType.wrap(pojoMember.getType(), typeMappings);
    return new JavaPojoMember(
        JavaName.fromName(pojoMember.getName()),
        pojoMember.getDescription(),
        javaType,
        pojoMember.getNecessity(),
        pojoMember.getNullability(),
        MemberType.OBJECT_MEMBER);
  }

  public JavaName getName() {
    return name;
  }

  public Nullability getNullability() {
    return nullability;
  }

  public Necessity getNecessity() {
    return necessity;
  }

  public String getDescription() {
    return description;
  }

  public JavaType getJavaType() {
    return javaType;
  }

  public MemberType getType() {
    return type;
  }

  public JavaPojoMember asObjectMember() {
    return JavaPojoMemberBuilder.create()
        .name(name)
        .description(description)
        .javaType(javaType)
        .necessity(necessity)
        .nullability(nullability)
        .type(MemberType.OBJECT_MEMBER)
        .andAllOptionals()
        .build();
  }

  public JavaPojoMember asAllOfMember() {
    return JavaPojoMemberBuilder.create()
        .name(name)
        .description(description)
        .javaType(javaType)
        .necessity(necessity)
        .nullability(nullability)
        .type(MemberType.ALL_OF_MEMBER)
        .andAllOptionals()
        .build();
  }

  public JavaPojoMember asOneOfMember() {
    return JavaPojoMemberBuilder.create()
        .name(name)
        .description(description)
        .javaType(javaType)
        .necessity(necessity)
        .nullability(nullability)
        .type(MemberType.ONE_OF_MEMBER)
        .andAllOptionals()
        .build();
  }

  public JavaPojoMember asAnyOfMember() {
    return JavaPojoMemberBuilder.create()
        .name(name)
        .description(description)
        .javaType(javaType)
        .necessity(necessity)
        .nullability(nullability)
        .type(MemberType.ANY_OF_MEMBER)
        .andAllOptionals()
        .build();
  }

  public boolean isOptional() {
    return necessity.isOptional();
  }

  public boolean isRequired() {
    return necessity.isRequired();
  }

  public boolean isNullable() {
    return nullability.isNullable();
  }

  public boolean isNotNullable() {
    return nullability.isNotNullable();
  }

  public boolean isRequiredAndNullable() {
    return isRequired() && isNullable();
  }

  public boolean isRequiredAndNotNullable() {
    return isRequired() && isNotNullable();
  }

  public boolean isOptionalAndNullable() {
    return isOptional() && isNullable();
  }

  public boolean isOptionalAndNotNullable() {
    return isOptional() && isNotNullable();
  }

  public JavaName getWitherName() {
    return prefixedMethodName("with");
  }

  public JavaName getIsPresentFlagName() {
    return IsPresentFlagName.fromName(name).getName();
  }

  public JavaName getIsNullFlagName() {
    return IsNullFlagName.fromName(name).getName();
  }

  public JavaName getGetterName() {
    return prefixedMethodName("get");
  }

  public JavaName getValidationGetterName(PojoSettings settings) {
    return JavaName.fromString(getGetterName().asString())
        .append(settings.getValidationMethods().getGetterSuffix());
  }

  public JavaName getGetterNameWithSuffix(PojoSettings settings) {
    return JavaName.fromString(getGetterName().asString()).append(determineSuffix(settings));
  }

  private String determineSuffix(PojoSettings settings) {
    final GetterSuffixes getterSuffixes = settings.getGetterSuffixes();
    if (isRequiredAndNotNullable()) {
      return getterSuffixes.getRequiredSuffix();
    } else if (isRequiredAndNullable()) {
      return getterSuffixes.getRequiredNullableSuffix();
    } else if (isOptionalAndNotNullable()) {
      return getterSuffixes.getOptionalSuffix();
    } else {
      return getterSuffixes.getOptionalNullableSuffix();
    }
  }

  public JavaName prefixedMethodName(String prefix) {
    return name.prefixedMethodName(prefix);
  }

  public PList<TechnicalPojoMember> getTechnicalMembers() {
    final TechnicalPojoMember technicalPojoMember = TechnicalPojoMember.wrapPojoMember(this);
    if (isRequiredAndNullable()) {
      return PList.of(
          technicalPojoMember, TechnicalPojoMember.isPresentFlagMember(getIsPresentFlagName()));
    } else if (isOptionalAndNullable()) {
      return PList.of(
          technicalPojoMember, TechnicalPojoMember.isNullFlagMember(getIsNullFlagName()));
    } else {
      return PList.single(technicalPojoMember);
    }
  }

  public String tristateToProperty() {
    return TRISTATE_TO_PROPERTY;
  }

  public String tristateToIsNullFlag() {
    return TRISTATE_TO_ISNULL_FLAG;
  }

  /** Creates {@link EnumContent} for this member in case its type is an {@link EnumType}. */
  public Optional<EnumContent> asEnumContent() {
    final Function<JavaEnumType, Optional<EnumContent>> toEnumPojo =
        enumType ->
            Optional.of(
                EnumContentBuilder.create()
                    .className(JavaName.fromName(enumType.getQualifiedClassName().getClassName()))
                    .description(getDescription())
                    .members(enumType.getMembers())
                    .build());
    return javaType.fold(
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        toEnumPojo,
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        ignore -> Optional.empty());
  }

  /**
   * In case this member is an enum type, it converts the classname of the enum so that it's
   * referenced via outer-class {@code javaPojoName}.
   */
  public JavaPojoMember asInnerEnumOf(JavaName javaPojoName) {
    final JavaType newType =
        javaType.fold(
            arrayType -> arrayType,
            booleanType -> booleanType,
            enumType -> enumType.asInnerClassOf(javaPojoName),
            mapType -> mapType,
            noType -> noType,
            numericType -> numericType,
            integerType -> integerType,
            objectType -> objectType,
            stringType -> stringType);
    return new JavaPojoMember(name, description, newType, necessity, nullability, type);
  }

  public enum MemberType {
    OBJECT_MEMBER(false),
    ALL_OF_MEMBER(true),
    ONE_OF_MEMBER(true),
    ANY_OF_MEMBER(true),
    ARRAY_VALUE(false);

    private final boolean isComposedMember;

    MemberType(boolean isComposedMember) {
      this.isComposedMember = isComposedMember;
    }

    public boolean isComposedMember() {
      return isComposedMember;
    }
  }
}
