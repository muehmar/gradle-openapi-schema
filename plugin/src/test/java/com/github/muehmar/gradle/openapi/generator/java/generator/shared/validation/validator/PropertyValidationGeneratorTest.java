package com.github.muehmar.gradle.openapi.generator.java.generator.shared.validation.validator;

import static com.github.muehmar.gradle.openapi.generator.java.generator.shared.validation.validator.PropertyValidationGenerator.memberValidationGenerator;
import static com.github.muehmar.gradle.openapi.generator.java.model.JavaPojoMembers.list;
import static com.github.muehmar.gradle.openapi.generator.java.model.JavaPojoMembers.map;
import static com.github.muehmar.gradle.openapi.generator.java.model.JavaPojoMembers.requiredDouble;
import static com.github.muehmar.gradle.openapi.generator.java.model.JavaPojoMembers.requiredInteger;
import static com.github.muehmar.gradle.openapi.generator.java.model.JavaPojoMembers.requiredString;
import static com.github.muehmar.gradle.openapi.generator.model.Necessity.REQUIRED;
import static com.github.muehmar.gradle.openapi.generator.model.Nullability.NOT_NULLABLE;
import static com.github.muehmar.gradle.openapi.generator.model.Nullability.NULLABLE;
import static com.github.muehmar.gradle.openapi.generator.settings.TestPojoSettings.defaultTestSettings;
import static io.github.muehmar.codegenerator.writer.Writer.javaWriter;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.annotations.SnapshotName;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaPojoMember;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaPojoMembers;
import com.github.muehmar.gradle.openapi.generator.java.model.name.QualifiedClassName;
import com.github.muehmar.gradle.openapi.generator.java.model.type.JavaNumericType;
import com.github.muehmar.gradle.openapi.generator.java.model.type.JavaStringType;
import com.github.muehmar.gradle.openapi.generator.model.constraints.Constraints;
import com.github.muehmar.gradle.openapi.generator.model.constraints.DecimalMax;
import com.github.muehmar.gradle.openapi.generator.model.constraints.DecimalMin;
import com.github.muehmar.gradle.openapi.generator.model.constraints.MultipleOf;
import com.github.muehmar.gradle.openapi.generator.model.constraints.PropertyCount;
import com.github.muehmar.gradle.openapi.generator.model.constraints.Size;
import com.github.muehmar.gradle.openapi.generator.model.name.PojoName;
import com.github.muehmar.gradle.openapi.generator.model.type.ArrayType;
import com.github.muehmar.gradle.openapi.generator.model.type.NumericType;
import com.github.muehmar.gradle.openapi.generator.model.type.ObjectType;
import com.github.muehmar.gradle.openapi.generator.model.type.StringType;
import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import com.github.muehmar.gradle.openapi.generator.settings.TypeMappings;
import com.github.muehmar.gradle.openapi.snapshot.SnapshotTest;
import com.github.muehmar.gradle.openapi.snapshot.SnapshotUtil;
import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.codegenerator.writer.Writer;
import java.math.BigDecimal;
import org.junit.jupiter.api.Test;

@SnapshotTest
class PropertyValidationGeneratorTest {
  private Expect expect;

  @Test
  @SnapshotName("requiredStringWithoutCondition")
  void generate_when_requiredStringWithoutCondition_then_matchSnapshot() {
    final Generator<JavaPojoMember, PojoSettings> generator = memberValidationGenerator();

    final JavaPojoMember stringType =
        requiredString()
            .withNecessity(REQUIRED)
            .withNullability(NOT_NULLABLE)
            .withJavaType(
                JavaStringType.wrap(
                    StringType.noFormat().withConstraints(Constraints.empty()),
                    TypeMappings.empty()));

    final Writer writer = generator.generate(stringType, defaultTestSettings(), javaWriter());

    expect.toMatchSnapshot(SnapshotUtil.writerSnapshot(writer));
  }

  @Test
  @SnapshotName("requiredStringWithCondition")
  void generate_when_requiredStringWithCondition_then_matchSnapshot() {
    final Generator<JavaPojoMember, PojoSettings> generator = memberValidationGenerator();

    final JavaPojoMember stringType =
        requiredString()
            .withNecessity(REQUIRED)
            .withNullability(NOT_NULLABLE)
            .withJavaType(
                JavaStringType.wrap(
                    StringType.noFormat().withConstraints(Constraints.ofSize(Size.ofMin(5))),
                    TypeMappings.empty()));

    final Writer writer = generator.generate(stringType, defaultTestSettings(), javaWriter());

    expect.toMatchSnapshot(SnapshotUtil.writerSnapshot(writer));
  }

  @Test
  @SnapshotName("requiredNullableStringWithoutCondition")
  void generate_when_requiredNullableStringWithoutCondition_then_matchSnapshot() {
    final Generator<JavaPojoMember, PojoSettings> generator = memberValidationGenerator();

    final JavaPojoMember stringType =
        requiredString()
            .withNecessity(REQUIRED)
            .withNullability(NULLABLE)
            .withJavaType(
                JavaStringType.wrap(
                    StringType.noFormat().withConstraints(Constraints.empty()),
                    TypeMappings.empty()));

    final Writer writer = generator.generate(stringType, defaultTestSettings(), javaWriter());

    expect.toMatchSnapshot(SnapshotUtil.writerSnapshot(writer));
  }

  @Test
  @SnapshotName("requiredNullableStringWithCondition")
  void generate_when_requiredNullableStringWithCondition_then_matchSnapshot() {
    final Generator<JavaPojoMember, PojoSettings> generator = memberValidationGenerator();

    final JavaPojoMember stringType =
        requiredString()
            .withNecessity(REQUIRED)
            .withNullability(NULLABLE)
            .withJavaType(
                JavaStringType.wrap(
                    StringType.noFormat().withConstraints(Constraints.ofSize(Size.ofMin(5))),
                    TypeMappings.empty()));

    final Writer writer = generator.generate(stringType, defaultTestSettings(), javaWriter());

    expect.toMatchSnapshot(SnapshotUtil.writerSnapshot(writer));
  }

  @Test
  @SnapshotName("integerWithMinAndMax")
  void generate_when_integerWithMinAndMax_then_matchSnapshot() {
    final Generator<JavaPojoMember, PojoSettings> generator = memberValidationGenerator();

    final Writer writer =
        generator.generate(requiredInteger(), defaultTestSettings(), javaWriter());

    expect.toMatchSnapshot(SnapshotUtil.writerSnapshot(writer));
  }

  @Test
  @SnapshotName("stringWithSize")
  void generate_when_stringWithMinAndMaxLength_then_matchSnapshot() {
    final Generator<JavaPojoMember, PojoSettings> generator = memberValidationGenerator();

    final JavaPojoMember stringType =
        requiredString()
            .withJavaType(
                JavaStringType.wrap(
                    StringType.noFormat().withConstraints(Constraints.ofSize(Size.of(10, 50))),
                    TypeMappings.empty()));

    final Writer writer = generator.generate(stringType, defaultTestSettings(), javaWriter());

    expect.toMatchSnapshot(SnapshotUtil.writerSnapshot(writer));
  }

  @Test
  @SnapshotName("listWithSize")
  void generate_when_listWithMinAndMaxLength_then_matchSnapshot() {
    final Generator<JavaPojoMember, PojoSettings> generator = memberValidationGenerator();

    final JavaPojoMember listType =
        list(StringType.noFormat(), REQUIRED, NOT_NULLABLE, Constraints.ofSize(Size.of(10, 50)));

    final Writer writer = generator.generate(listType, defaultTestSettings(), javaWriter());

    expect.toMatchSnapshot(SnapshotUtil.writerSnapshot(writer));
  }

  @Test
  @SnapshotName("listUniqueItems")
  void generate_when_listUniqueItems_then_matchSnapshot() {
    final Generator<JavaPojoMember, PojoSettings> generator = memberValidationGenerator();

    final JavaPojoMember listType =
        list(StringType.noFormat(), REQUIRED, NOT_NULLABLE, Constraints.ofUniqueItems(true));

    final Writer writer = generator.generate(listType, defaultTestSettings(), javaWriter());

    expect.toMatchSnapshot(SnapshotUtil.writerSnapshot(writer));
  }

  @Test
  @SnapshotName("mapWithSize")
  void generate_when_mapWithMinAndMaxLength_then_matchSnapshot() {
    final Generator<JavaPojoMember, PojoSettings> generator = memberValidationGenerator();

    final JavaPojoMember listType =
        map(
            StringType.noFormat(),
            StringType.noFormat(),
            REQUIRED,
            NOT_NULLABLE,
            Constraints.ofSize(Size.of(10, 50)));

    final Writer writer = generator.generate(listType, defaultTestSettings(), javaWriter());

    expect.toMatchSnapshot(SnapshotUtil.writerSnapshot(writer));
  }

  @Test
  @SnapshotName("mapWithPropertyCount")
  void generate_when_mapWithPropertyCount_then_matchSnapshot() {
    final Generator<JavaPojoMember, PojoSettings> generator = memberValidationGenerator();

    final JavaPojoMember listType =
        map(
            StringType.noFormat(),
            StringType.noFormat(),
            REQUIRED,
            NOT_NULLABLE,
            Constraints.ofPropertiesCount(PropertyCount.ofMinAndMaxProperties(5, 10)));

    final Writer writer = generator.generate(listType, defaultTestSettings(), javaWriter());

    expect.toMatchSnapshot(SnapshotUtil.writerSnapshot(writer));
  }

  @Test
  @SnapshotName("byteArraySize")
  void generate_when_byteArrayWithMinAndMaxLength_then_matchSnapshot() {
    final Generator<JavaPojoMember, PojoSettings> generator = memberValidationGenerator();

    final JavaPojoMember byteArrayType =
        requiredString()
            .withJavaType(
                JavaStringType.wrap(
                    StringType.ofFormat(StringType.Format.BINARY)
                        .withConstraints(Constraints.ofSize(Size.of(10, 50))),
                    TypeMappings.empty()));

    final Writer writer = generator.generate(byteArrayType, defaultTestSettings(), javaWriter());

    expect.toMatchSnapshot(SnapshotUtil.writerSnapshot(writer));
  }

  @Test
  @SnapshotName("doubleWithDecimalMinMaxExclusive")
  void generate_when_doubleWithDecimalMinMaxExclusive_then_matchSnapshot() {
    final Generator<JavaPojoMember, PojoSettings> generator = memberValidationGenerator();

    final JavaPojoMember doubleMember =
        requiredDouble()
            .withJavaType(
                JavaNumericType.wrap(
                    NumericType.formatDouble()
                        .withConstraints(
                            Constraints.ofDecimalMinAndMax(
                                new DecimalMin("50.1", false), new DecimalMax("100.1", false))),
                    TypeMappings.empty()));

    final Writer writer = generator.generate(doubleMember, defaultTestSettings(), javaWriter());

    expect.toMatchSnapshot(SnapshotUtil.writerSnapshot(writer));
  }

  @Test
  @SnapshotName("doubleWithDecimalMinMaxInclusive")
  void generate_when_doubleWithDecimalMinMaxInclusive_then_matchSnapshot() {
    final Generator<JavaPojoMember, PojoSettings> generator = memberValidationGenerator();

    final JavaPojoMember doubleMember =
        requiredDouble()
            .withJavaType(
                JavaNumericType.wrap(
                    NumericType.formatDouble()
                        .withConstraints(
                            Constraints.ofDecimalMinAndMax(
                                new DecimalMin("50.1", true), new DecimalMax("100.1", true))),
                    TypeMappings.empty()));

    final Writer writer = generator.generate(doubleMember, defaultTestSettings(), javaWriter());

    expect.toMatchSnapshot(SnapshotUtil.writerSnapshot(writer));
  }

  @Test
  @SnapshotName("stringWithPattern")
  void generate_when_stringWithPattern_then_matchSnapshot() {
    final Generator<JavaPojoMember, PojoSettings> generator = memberValidationGenerator();

    final Writer writer = generator.generate(requiredString(), defaultTestSettings(), javaWriter());

    expect.toMatchSnapshot(SnapshotUtil.writerSnapshot(writer));
  }

  @Test
  @SnapshotName("stringWithEmail")
  void generate_when_stringWithEmail_then_matchSnapshot() {
    final Generator<JavaPojoMember, PojoSettings> generator = memberValidationGenerator();

    final Writer writer =
        generator.generate(
            requiredString()
                .withJavaType(
                    JavaStringType.wrap(
                        StringType.noFormat().withConstraints(Constraints.ofEmail()),
                        TypeMappings.empty())),
            defaultTestSettings(),
            javaWriter());

    expect.toMatchSnapshot(SnapshotUtil.writerSnapshot(writer));
  }

  @Test
  @SnapshotName("mapWithListValueType")
  void generate_when_mapWithListValueType_then_matchSnapshot() {
    final Generator<JavaPojoMember, PojoSettings> generator = memberValidationGenerator();

    final StringType stringType =
        StringType.noFormat().withConstraints(Constraints.ofSize(Size.ofMax(50)));
    final ArrayType listType =
        ArrayType.ofItemType(stringType).withConstraints(Constraints.ofSize(Size.ofMin(8)));
    final JavaPojoMember mapType =
        map(
            StringType.noFormat(),
            listType,
            REQUIRED,
            NOT_NULLABLE,
            Constraints.ofSize(Size.of(10, 50)));

    final Writer writer = generator.generate(mapType, defaultTestSettings(), javaWriter());

    expect.toMatchSnapshot(SnapshotUtil.writerSnapshot(writer));
  }

  @Test
  @SnapshotName("customObjectType")
  void generate_when_customObjectType_then_matchSnapshot() {
    final Generator<JavaPojoMember, PojoSettings> generator = memberValidationGenerator();

    final JavaPojoMember objectMember =
        JavaPojoMembers.object(QualifiedClassName.ofName("CustomObject"));

    final Writer writer = generator.generate(objectMember, defaultTestSettings(), javaWriter());

    expect.toMatchSnapshot(SnapshotUtil.writerSnapshot(writer));
  }

  @Test
  @SnapshotName("openapiObjectType")
  void generate_when_openapiObjectType_then_matchSnapshot() {
    final Generator<JavaPojoMember, PojoSettings> generator = memberValidationGenerator();

    final JavaPojoMember objectMember =
        JavaPojoMembers.object(ObjectType.ofName(PojoName.ofNameAndSuffix("OpenapiObject", "Dto")));

    final Writer writer = generator.generate(objectMember, defaultTestSettings(), javaWriter());

    expect.toMatchSnapshot(SnapshotUtil.writerSnapshot(writer));
  }

  @Test
  @SnapshotName("doubleMultipleOf")
  void generate_when_doubleMultipleOf_then_matchSnapshot() {
    final Generator<JavaPojoMember, PojoSettings> generator = memberValidationGenerator();

    final JavaPojoMember doubleMember =
        requiredDouble()
            .withJavaType(
                JavaNumericType.wrap(
                    NumericType.formatDouble()
                        .withConstraints(
                            Constraints.ofMultipleOf(new MultipleOf(new BigDecimal("129")))),
                    TypeMappings.empty()));

    final Writer writer = generator.generate(doubleMember, defaultTestSettings(), javaWriter());

    expect.toMatchSnapshot(SnapshotUtil.writerSnapshot(writer));
  }
}
