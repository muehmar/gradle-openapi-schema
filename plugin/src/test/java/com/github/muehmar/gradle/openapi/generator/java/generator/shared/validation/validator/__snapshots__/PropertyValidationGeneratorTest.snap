byteArraySize=[
.
.
private boolean isStringValValid() {
  if(stringVal != null) {
    return 10 <= stringVal.length
        && stringVal.length <= 50;
  }

  return false;
}
]


customObjectType=[
.
.
private boolean isObjectValid() {
  return object != null;
}
]


doubleMultipleOf=[
.
.
private boolean isDoubleValValid() {
  if(doubleVal != null) {
    return isDoubleValMultipleOfValid();
  }

  return false;
}
]


doubleWithDecimalMinMaxExclusive=[
java.math.BigDecimal

private boolean isDoubleValValid() {
  if(doubleVal != null) {
    return 0 < BigDecimal.valueOf(doubleVal).compareTo(new BigDecimal("50.1"))
        && BigDecimal.valueOf(doubleVal).compareTo(new BigDecimal("100.1")) < 0;
  }

  return false;
}
]


doubleWithDecimalMinMaxInclusive=[
java.math.BigDecimal

private boolean isDoubleValValid() {
  if(doubleVal != null) {
    return 0 <= BigDecimal.valueOf(doubleVal).compareTo(new BigDecimal("50.1"))
        && BigDecimal.valueOf(doubleVal).compareTo(new BigDecimal("100.1")) <= 0;
  }

  return false;
}
]


integerWithMinAndMax=[
.
.
private boolean isIntValValid() {
  if(intVal != null) {
    return 10L <= intVal
        && intVal <= 50L;
  }

  return false;
}
]


listNullableItems=[
.
.
private boolean isListValValid() {
  if(listVal != null) {
    return listVal.stream().allMatch(this::isListValValueValid);
  }

  return false;
}

private boolean isListValValueValid(String listValValue) {
  return true;
}
]


listUniqueItems=[
.
.
private boolean isListValValid() {
  if(listVal != null) {
    return hasListValUniqueItems()
        && listVal.stream().allMatch(this::isListValValueValid);
  }

  return false;
}

private boolean isListValValueValid(String listValValue) {
  return listValValue != null;
}
]


listWithSize=[
.
.
private boolean isListValValid() {
  if(listVal != null) {
    return 10 <= listVal.size()
        && listVal.size() <= 50
        && listVal.stream().allMatch(this::isListValValueValid);
  }

  return false;
}

private boolean isListValValueValid(String listValValue) {
  return listValValue != null;
}
]


listWithStringItemsWithConstraints=[
.
.
private boolean isListValValid() {
  if(listVal != null) {
    return listVal.stream().allMatch(this::isListValValueValid);
  }

  return false;
}

private boolean isListValValueValid(String listValValue) {
  if(listValValue != null) {
    return 10 <= listValValue.length()
        && listValValue.length() <= 50;
  }

  return false;
}
]


mapWithListValueType=[
.
.
private boolean isMapValValid() {
  if(mapVal != null) {
    return 10 <= mapVal.size()
        && mapVal.size() <= 50
        && mapVal.values().stream().allMatch(this::isMapValValueValid);
  }

  return false;
}

private boolean isMapValValueValid(List<String> mapValValue) {
  if(mapValValue != null) {
    return 8 <= mapValValue.size()
        && mapValValue.stream().allMatch(this::isMapValValueValueValid);
  }

  return false;
}

private boolean isMapValValueValueValid(String mapValValueValue) {
  if(mapValValueValue != null) {
    return mapValValueValue.length() <= 50;
  }

  return false;
}
]


mapWithNullableProperties=[
.
.
private boolean isMapValValid() {
  if(mapVal != null) {
    return 10 <= mapVal.size()
        && mapVal.size() <= 50
        && mapVal.values().stream().allMatch(this::isMapValValueValid);
  }

  return false;
}

private boolean isMapValValueValid(String mapValValue) {
  return true;
}
]


mapWithPropertyCount=[
.
.
private boolean isMapValValid() {
  if(mapVal != null) {
    return 5 <= mapVal.size()
        && mapVal.size() <= 10
        && mapVal.values().stream().allMatch(this::isMapValValueValid);
  }

  return false;
}

private boolean isMapValValueValid(String mapValValue) {
  return mapValValue != null;
}
]


mapWithSize=[
.
.
private boolean isMapValValid() {
  if(mapVal != null) {
    return 10 <= mapVal.size()
        && mapVal.size() <= 50
        && mapVal.values().stream().allMatch(this::isMapValValueValid);
  }

  return false;
}

private boolean isMapValValueValid(String mapValValue) {
  return mapValValue != null;
}
]


openapiObjectType=[
.
.
private boolean isObjectValid() {
  if(object != null) {
    return object.isValid();
  }

  return false;
}
]


optionalNotNullableStringWithCondition=[
.
.
private boolean isOptionalStringValValid() {
  if(optionalStringVal != null) {
    return 5 <= optionalStringVal.length();
  }

  return isOptionalStringValNotNull;
}
]


optionalNullableStringWithCondition=[
.
.
private boolean isOptionalStringValValid() {
  if(optionalStringVal != null) {
    return 5 <= optionalStringVal.length();
  }

  return true;
}
]


requiredNullableStringWithCondition=[
.
.
private boolean isStringValValid() {
  if(stringVal != null) {
    return 5 <= stringVal.length();
  }

  return isStringValPresent;
}
]


requiredNullableStringWithoutCondition=[
.
.
private boolean isStringValValid() {
  return (stringVal != null || isStringValPresent);
}
]


requiredStringWithCondition=[
.
.
private boolean isStringValValid() {
  if(stringVal != null) {
    return 5 <= stringVal.length();
  }

  return false;
}
]


requiredStringWithoutCondition=[
.
.
private boolean isStringValValid() {
  return stringVal != null;
}
]


stringWithEmail=[
com.github.muehmar.openapi.util.EmailValidator

private boolean isStringValValid() {
  if(stringVal != null) {
    return EmailValidator.isValid(stringVal);
  }

  return false;
}
]


stringWithPattern=[
.
.
private boolean isStringValValid() {
  if(stringVal != null) {
    return java.util.regex.Pattern.matches("Hello", stringVal);
  }

  return false;
}
]


stringWithSize=[
.
.
private boolean isStringValValid() {
  if(stringVal != null) {
    return 10 <= stringVal.length()
        && stringVal.length() <= 50;
  }

  return false;
}
]


unsupportedConstraintsForType=[
.
.
private boolean isUnsupportedValid() {
  return unsupported != null;
}
]